src/esp/gfx/ssao/HBAOHelper2.cpp

```diff
@@ -0,0 +1,787 @@
+      glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+      glBindTexture(GL_TEXTURE_2D, 0);
+    }
+
+    nvgl::newTexture(textures.hbao2_resultarray, GL_TEXTURE_2D_ARRAY);
+    glBindTexture(GL_TEXTURE_2D_ARRAY, textures.hbao2_resultarray);
+    glTexStorage3D(GL_TEXTURE_2D_ARRAY, 1, formatAO, quarterWidth, quarterHeight,
+                  HBAO_RANDOM_ELEMENTS);
+    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glBindTexture(GL_TEXTURE_2D_ARRAY, 0);
+
+    GLenum drawbuffers[NUM_MRT];
+    for (int layer = 0; layer < NUM_MRT; layer++) {
+      drawbuffers[layer] = GL_COLOR_ATTACHMENT0 + layer;
+    }
+
+    nvgl::newFramebuffer(fbos.hbao2_deinterleave);
+    glBindFramebuffer(GL_FRAMEBUFFER, fbos.hbao2_deinterleave);
+    glDrawBuffers(NUM_MRT, drawbuffers);
+    glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+    nvgl::newFramebuffer(fbos.hbao2_calc);
+    glBindFramebuffer(GL_FRAMEBUFFER, fbos.hbao2_calc);
+#if USE_AO_LAYERED_SINGLEPASS == AO_LAYERED_IMAGE
+    // this fbo will not have any attachments and therefore requires rasterizer to
+    // be configured through default parameters
+    glFramebufferParameteri(GL_FRAMEBUFFER, GL_FRAMEBUFFER_DEFAULT_WIDTH,
+                            quarterWidth);
+    glFramebufferParameteri(GL_FRAMEBUFFER, GL_FRAMEBUFFER_DEFAULT_HEIGHT,
+                            quarterHeight);
+#endif
+#if USE_AO_LAYERED_SINGLEPASS == AO_LAYERED_GS
+    glFramebufferTexture(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0,
+                        textures.hbao2_resultarray, 0);
+#endif
+
+    glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+#endif
+
+    return true;
+  }
+
+  void drawLinearDepth(float projNearPlane,
+                       float projFarPlane,
+                       bool isProjOrtho,
+                       int width,
+                       int height,
+                       int sampleIdx) {
+    // temp resetState required because we're mixing Magnum OpenGL and external
+    // OpenGL.
+    Mn::GL::Context::current().resetState();
+
+    // glBindFramebuffer(GL_FRAMEBUFFER, fbos.depthlinear);
+    //  todo
+    //  setViewport() doesn't need to be called if a correct viewport was passed
+    //  to hbao2CalcFramebuffer constructor already
+    //.setViewport({{}, {quarterWidth, quarterHeight}})
+    fbos.depthlinear.bind();
+
+    if (m_config.samples > 1) {
+      // glUseProgram(m_progManager.get(programs.depth_linearize_msaa));
+      // glUniform4f(0, projNearPlane * projFarPlane,
+      //             projNearPlane - projFarPlane, projFarPlane,
+      //             isProjOrtho ? 0.0f : 1.0f);
+      // glUniform1i(1, sampleIdx);
+      programs.depth_linearize_msaa.setUniform(
```

Comment:
@mosra is there a portability concern about the hard-coded `0` here corresponding to the `layout(location=0) uniform vec4 clipInfo` in `depthlinearize.frag.glsl`?

Reply:
So I'm imagining adding a new version of `setUniform` to `HBAOShader` that also takes the string name ("clipInfo"), and internally looks up and caches the location. (A design goal of `HBAOShader` here is to support more-or-less line-by-line porting from raw OpenGL, hence my preference to do the lookat at this callsite rather than at shader-creation time.)

Related: happy to just have the entire HBAO effect compiled out and unsupported in WebGL for our initial PR.